File: test/verify/fixture.py
"""Python mirror of test/verify/fixture.ts: the same cases through the Python package, so the two
packages' output can be diffed. Run from the repo root with the python venv:
    python/.venv/bin/python test/verify/fixture.py out.json
"""
import json
import sys

sys.path.insert(0, "python/src")
from shortcutkit import ACTIONS, PARAM_KINDS, PARAM_CHOICES, PARAM_VARIABLE_TYPES, Shortcut, actions, ref, text, variable, shortcut_input, clipboard, current_date, ask, device_details, current_app, repeat_item  # noqa: E402

encodings = json.load(open("data/parameter-encodings.json"))
DEFAULTS = encodings["actionDefaults"]


def runtime(identifier, key):
    return next((p for p in encodings["actionParameters"].get(identifier, []) if p["key"] == key), None)


def is_plain(v):
    return isinstance(v, (str, int, float, bool))


seed = Shortcut("Seed")
source = seed.action(actions.GETSTOREDCONTENT, WFStoredContentKey="seed")
REF = ref(source)
REFERENCES = [("ref", REF), ("shortcutInput", shortcut_input()), ("variable", variable("MyVar")), ("clipboard", clipboard()), ("currentDate", current_date()), ("ask", ask("Which?")), ("deviceDetails", device_details()), ("currentApp", current_app())]


def samples(identifier, key, kind):
    choices = PARAM_CHOICES.get(identifier, {}).get(key, [])
    rt = runtime(identifier, key) or {}
    candidates = [rt.get("plainSample"), DEFAULTS.get(identifier, {}).get(key)]

    def same_type(v, fallback):
        if isinstance(fallback, bool):
            return isinstance(v, bool)
        if isinstance(fallback, (int, float)):
            return isinstance(v, (int, float)) and not isinstance(v, bool)
        return isinstance(v, str)

    def plain(fallback):
        return next((v for v in candidates if v is not None and same_type(v, fallback)), fallback)
    return {
        "bool": [("bool", plain(True)), *REFERENCES],
        "number": [("number", plain(5)), *REFERENCES],
        "string": [("string", plain(choices[0] if choices else "hello")), *REFERENCES],
        "text": [("string", plain("hello")), ("text+ref", text("Hi ", REF)), *REFERENCES],
        "picker": [r for r in REFERENCES if r[0] != "ask"],
        "subject": [("ref", REF)],
        "plainString": [("string", "hello")],
        "plainNumber": [("number", 5)],
    }.get(kind, [])


LEGACY = {
    "is.workflow.actions.conditional": ["WFInput", "WFCondition", "WFConditionalActionString", "WFNumberValue", "WFConditionalLegacyComparisonBehavior", "WFEnumerationValue", "WFBooleanValue", "WFDate", "WFAnotherDate", "WFDuration"],
    "is.workflow.actions.choosefrommenu": ["WFMenuPrompt", "WFMenuItems", "WFMenuItemTitle", "WFMenuItemAttributedTitle"],
}
cases = []
for identifier in ACTIONS:
    for key, kind in PARAM_KINDS.get(identifier, {}).items():
        if key in LEGACY.get(identifier, []):
            continue
        allowed = PARAM_VARIABLE_TYPES.get(identifier, {}).get(key)
        for form, value in samples(identifier, key, kind):
            is_ref = isinstance(value, dict) and value.get("WFSerializationType") == "WFTextTokenAttachment"
            if allowed is not None and is_ref and value["Value"]["Type"] not in allowed:
                cases.append({"identifier": identifier, "key": key, "kind": kind, "form": form, "params": {key: value}, "expect": "rejected"})
                continue
            if allowed is not None and form == "text+ref":
                continue
            s = Shortcut("Case")
            a = s.action(identifier, **{key: value})
            params = {k: v for k, v in a["WFWorkflowActionParameters"].items() if k != "UUID"}
            rt = runtime(identifier, key)
            expect = "deviceValidated" if is_plain(value) and rt is not None and rt.get("readsPlain") is False else "read"
            cases.append({"identifier": identifier, "key": key, "kind": kind, "form": form, "params": params, "expect": expect})

s = Shortcut("Blocks", color="Teal", types=["ActionExtension"], input_classes=["WFSafariWebPageContentItem", "WFURLContentItem"])
got = s.action(actions.GETSTOREDCONTENT, WFStoredContentKey="k")
c = s.if_(ref(got), "has_any_value")
loop = s.repeat_each(ref(got))
s.action(actions.DOWNLOADURL, WFURL=repeat_item())
results = s.end_repeat_each(loop)
s.action(actions.PREVIEWDOCUMENT, WFInput=ref(results))
s.otherwise(c)
n = s.repeat_count(3)
s.action(actions.SHOWRESULT, Text=text("tick ", repeat_item()))
s.end_repeat_count(n)
s.end_if(c)
c2 = s.if_(ref(got), "is", "hello"); s.action(actions.SHOWRESULT, Text=text("is")); s.end_if(c2)
c3 = s.if_(ref(got), "less_than", 5); s.action(actions.SHOWRESULT, Text=text("lt")); s.end_if(c3)
c4 = s.if_(shortcut_input(), "contains", "x"); s.action(actions.SHOWRESULT, Text=text("has")); s.otherwise(c4); s.end_if(c4)
m = s.choose_from_menu("Pick", ["A", "B"])
s.menu_item(m, "A"); s.action(actions.SHOWRESULT, Text=ref(got))
s.menu_item(m, "B"); s.action(actions.SHOWRESULT, Text=shortcut_input())
s.end_menu(m)
s.action(actions.NOTES_CREATE_NOTE, name="n", content=text("c ", ref(got)))
shortcuts = [{"name": "Blocks", "plist": s.to_plist(), "actionCount": len(s.actions)}]

json.dump({"cases": cases, "shortcuts": shortcuts}, open(sys.argv[1], "w"), indent=1, ensure_ascii=False)
print(len(cases), "parameter cases across", len({c["identifier"] for c in cases}), "actions;", len(shortcuts), "whole shortcuts ->", sys.argv[1])

File: test/verify/fixture.ts
// Drives the library over every catalogue action and every parameter key with every value form
// the types allow, plus whole shortcuts built through the block helpers, and writes what the
// library produced. tools/verify-library-output.js then loads all of it through the engine.
//   bun test/verify/fixture.ts out.json
import { Shortcut, actions, ACTIONS, PARAM_KINDS, PARAM_CHOICES, PARAM_VARIABLE_TYPES, ref, text, variable, shortcutInput, clipboard, currentDate, ask, deviceDetails, currentApp, repeatItem } from "../../src/index";
import encodings from "../../data/parameter-encodings.json";

// expect: "read" (the library wrote it, the engine must read it) or "rejected" (the library refuses
// this form for the key, written raw here; the engine must not read it either).
// "deviceValidated": the engine read no plain sample for this key when the data was extracted (it checks the
// value against the device: a locale, a HomeKit home, a contacts group), so a plain value cannot be proven here.
type Case = { identifier: string; key: string; kind: string; form: string; params: Record<string, unknown>; expect: "read" | "rejected" | "deviceValidated" };
type RuntimeParam = { key: string; class: string; reads?: string[]; readsPlain?: boolean; plainSample?: string | number | boolean };
const runtime = (identifier: string, key: string): RuntimeParam | undefined => (encodings.actionParameters as Record<string, RuntimeParam[]>)[identifier]?.find((p) => p.key === key);
const defaults = encodings.actionDefaults as Record<string, Record<string, unknown>>;
const isPlain = (v: unknown): v is string | number | boolean => typeof v === "string" || typeof v === "number" || typeof v === "boolean";
const seed = new Shortcut("Seed");
const source = seed.action(actions.getstoredcontent, { WFStoredContentKey: "seed" });
const REF = ref(source);
// Every reference the library can write, by the variable type the engine names it.
const REFERENCES: Array<[string, unknown]> = [["ref", REF], ["shortcutInput", shortcutInput()], ["variable", variable("MyVar")], ["clipboard", clipboard()], ["currentDate", currentDate()], ["ask", ask("Which?")], ["deviceDetails", deviceDetails()], ["currentApp", currentApp()]];
const typeOf = (v: unknown): string => String((v as { Value: { Type: string } }).Value.Type);

function samples(identifier: string, key: string, kind: string): Array<[string, unknown]> {
  const choices = (PARAM_CHOICES as Record<string, Record<string, readonly string[]>>)[identifier]?.[key] ?? [];
  // The plain sample: the one the probe loaded (the key's default, else its first possible state), else the engine
  // default, else a choice, else a stand-in.
  const probed = runtime(identifier, key)?.plainSample, dflt = defaults[identifier]?.[key];
  const plain = (fallback: string | number | boolean) => (isPlain(probed) && typeof probed === typeof fallback ? probed : isPlain(dflt) && typeof dflt === typeof fallback ? dflt : fallback);
  switch (kind) {
    case "bool": return [["bool", plain(true)], ...REFERENCES];
    case "number": return [["number", plain(5)], ...REFERENCES];
    case "string": return [["string", plain(choices[0] ?? "hello")], ...REFERENCES];
    case "text": return [["string", plain("hello")], ["text+ref", text("Hi ", REF)], ...REFERENCES];
    case "picker": return REFERENCES.filter(([f]) => f !== "ask");
    case "subject": return [["ref", REF]];
    case "plainString": return [["string", "hello"]];
    case "plainNumber": return [["number", 5]];
    default: return []; // dictionary, quantity, filter, any: no single form is promised by the type
  }
}

// Legacy keys the engine only reads as a set (If's WFInput/WFCondition/…, Choose from Menu's);
// they are exercised through the helpers in the whole shortcuts below, not one key at a time.
const LEGACY: Record<string, string[]> = {
  "is.workflow.actions.conditional": ["WFInput", "WFCondition", "WFConditionalActionString", "WFNumberValue", "WFConditionalLegacyComparisonBehavior", "WFEnumerationValue", "WFBooleanValue", "WFDate", "WFAnotherDate", "WFDuration"],
  "is.workflow.actions.choosefrommenu": ["WFMenuPrompt", "WFMenuItems", "WFMenuItemTitle", "WFMenuItemAttributedTitle"],
};
const cases: Case[] = [];
for (const identifier of Object.keys(ACTIONS)) {
  const kinds = (PARAM_KINDS as Record<string, Record<string, string>>)[identifier] ?? {};
  for (const [key, kind] of Object.entries(kinds)) {
    if ((LEGACY[identifier] ?? []).includes(key)) continue;
    const allowed = (PARAM_VARIABLE_TYPES as Record<string, Record<string, readonly string[]>>)[identifier]?.[key];
    for (const [form, value] of samples(identifier, key, kind)) {
      const isRef = typeof value === "object" && value !== null && (value as { WFSerializationType?: string }).WFSerializationType === "WFTextTokenAttachment";
      if (allowed && isRef && !allowed.includes(typeOf(value))) {
        // The library refuses this; write it raw so the verifier can confirm the engine refuses it too.
        cases.push({ identifier, key, kind, form, params: { [key]: value }, expect: "rejected" });
        continue;
      }
      if (allowed && form === "text+ref") continue; // a token string embedding a reference is refused for the same reason
      const s = new Shortcut("Case");
      const a = s.action(identifier, { [key]: value } as never);
      const params = { ...a.WFWorkflowActionParameters }; delete (params as Record<string, unknown>).UUID;
      cases.push({ identifier, key, kind, form, params, expect: isPlain(value) && runtime(identifier, key)?.readsPlain === false ? "deviceValidated" : "read" });
    }
  }
}

// Whole shortcuts through the helpers: control flow, loop variables, share sheet, text wrapping.
const s = new Shortcut("Blocks", { color: "Teal", types: ["ActionExtension"], inputClasses: ["WFSafariWebPageContentItem", "WFURLContentItem"] });
const got = s.action(actions.getstoredcontent, { WFStoredContentKey: "k" });
const c = s.if(ref(got), "has_any_value");
const loop = s.repeatEach(ref(got));
s.action(actions.downloadurl, { WFURL: repeatItem() });
const results = s.endRepeatEach(loop);
s.action(actions.previewdocument, { WFInput: ref(results) });
s.otherwise(c);
const n = s.repeatCount(3);
s.action(actions.showresult, { Text: text("tick ", repeatItem()) });
s.endRepeatCount(n);
s.endIf(c);
// Every If form the helper can write.
const c2 = s.if(ref(got), "is", "hello"); s.action(actions.showresult, { Text: text("is") }); s.endIf(c2);
const c3 = s.if(ref(got), "less_than", 5); s.action(actions.showresult, { Text: text("lt") }); s.endIf(c3);
const c4 = s.if(shortcutInput(), "contains", "x"); s.action(actions.showresult, { Text: text("has") }); s.otherwise(c4); s.endIf(c4);
const m = s.chooseFromMenu("Pick", ["A", "B"]);
s.menuItem(m, "A"); s.action(actions.showresult, { Text: ref(got) });
s.menuItem(m, "B"); s.action(actions.showresult, { Text: shortcutInput() });
s.endMenu(m);
s.action(actions.notes_create_note, { name: "n", content: text("c ", ref(got)) });
const shortcuts = [{ name: "Blocks", plist: s.toPlist(), actionCount: s.actions.length }];

const out = process.argv[2];
if (!out) throw new Error("usage: bun test/verify/fixture.ts out.json");
await Bun.write(out, JSON.stringify({ cases, shortcuts }, null, 1));
console.log(cases.length, "parameter cases across", new Set(cases.map((x) => x.identifier)).size, "actions;", shortcuts.length, "whole shortcuts ->", out);
